File: tools/checksum.lua
-- Run in Studio (Edit). Same checksum as tools/checksum.py.
local paths = {
	"ReplicatedStorage.Shared.Config",
	"ReplicatedStorage.Shared.UI",
	"ServerScriptService.Server.Walkers",
	"ServerScriptService.Server.Remotes",
	"ServerScriptService.Server.DataService",
	"ServerScriptService.Server.StudentProps",
	"ServerScriptService.Server.StudentFactory",
	"ServerScriptService.Server.PlotService",
	"ServerScriptService.Server.HallService",
	"ServerScriptService.Server.DebugBridge",
	"ServerScriptService.Server.StealService",
	"ServerScriptService.Server.SchoolService",
	"ServerScriptService.Server.Main",
	"StarterPlayer.StarterPlayerScripts.HUD",
	"StarterPlayer.StarterPlayerScripts.Menus",
	"StarterPlayer.StarterPlayerScripts.Prompts",
	"StarterPlayer.StarterPlayerScripts.Effects",
}
local out = {}
for _, p in paths do
	local inst = game
	for part in p:gmatch("[^%.]+") do
		inst = inst and inst:FindFirstChild(part)
	end
	if inst then
		local src = inst.Source:gsub("\r\n", "\n"):gsub("\n+$", "")
		local s = 0
		for i = 1, #src do
			s = (s + string.byte(src, i) * ((i - 1) % 251 + 1)) % 2147483648
		end
		table.insert(out, ("%s %d %d"):format(p, #src, s))
	end
end
return table.concat(out, "\n")

File: tools/checksum.py
# Prints a checksum per source file, matching tools/checksum.lua and tools/push.lua in Studio.
# checksum = (len, sum over bytes of b * (i % 251 + 1)) mod 2^31, CRLF normalised to LF,
# trailing newlines stripped (Studio drops them).
#   python tools/checksum.py            every file under src/
#   python tools/checksum.py Audio HUD  only paths containing one of the words
import sys, pathlib

BASES = {
    "Shared": "ReplicatedStorage.Shared",
    "Server": "ServerScriptService.Server",
    "Client": "StarterPlayer.StarterPlayerScripts",
    "First": "ReplicatedFirst",
}

root = pathlib.Path(__file__).resolve().parent.parent / "src"
words = sys.argv[1:]
for folder, base in BASES.items():
    for f in sorted((root / folder).glob("*.lua")):
        name = f.name[:-4]
        for suffix in (".server", ".client"):
            if name.endswith(suffix):
                name = name[: -len(suffix)]
        path = f"{base}.{name}"
        if words and not any(w in path for w in words):
            continue
        data = f.read_bytes().replace(b"\r\n", b"\n").rstrip(b"\n")
        s = 0
        for i, b in enumerate(data):
            s = (s + b * (i % 251 + 1)) % 2147483648
        print(f"{path} {len(data)} {s}")
